controllers/controller_ref_manager.go

```diff
@@ -434,29 +346,74 @@ func (m *EtcdDruidRefManager) ClaimConfigMaps(cms *corev1.ConfigMapList, filters
 	return claimed, utilerrors.NewAggregate(errlist)
 }
 
-// AdoptConfigMap sends a patch to take control of the Etcd. It returns the error if
+// AdoptResource sends a patch to take control of the Etcd. It returns the error if
 // the patching fails.
-func (m *EtcdDruidRefManager) AdoptConfigMap(cm *corev1.ConfigMap) error {
+func (m *EtcdDruidRefManager) AdoptResource(obj metav1.Object) error {
 	if err := m.CanAdopt(); err != nil {
-		return fmt.Errorf("can't adopt configmap %v/%v (%v): %v", cm.Namespace, cm.Name, cm.UID, err)
+		return fmt.Errorf("can't adopt resource %v/%v (%v): %v", obj.GetNamespace(), obj.GetName(), obj.GetUID(), err)
+	}
+
+	var clone metav1.Object
+	switch objType := obj.(type) {
+	case *appsv1.StatefulSet:
+		clone = obj.(*appsv1.StatefulSet).DeepCopy()
+		owners := clone.GetOwnerReferences()
+		ownersCopy := []metav1.OwnerReference{}
+		for i := range owners {
+			owner := &owners[i]
+			if owner.UID == m.Controller.GetUID() {
+				continue
+			}
+			ownersCopy = append(ownersCopy, *owner)
+		}
+		if len(ownersCopy) == 0 {
+			clone.SetOwnerReferences(nil)
+		} else {
+			clone.SetOwnerReferences(ownersCopy)
+		}
+	case *corev1.ConfigMap:
+		clone = obj.(*corev1.ConfigMap).DeepCopy()
+		// Note that ValidateOwnerReferences() will reject this patch if another
+		// OwnerReference exists with controller=true.
+		if err := controllerutil.SetControllerReference(m.Controller, clone, m.reconciler.Scheme); err != nil {
+			return err
+		}
+	case *corev1.Service:
+		clone = obj.(*corev1.Service).DeepCopy()
+		// Note that ValidateOwnerReferences() will reject this patch if another
+		// OwnerReference exists with controller=true.
+		if err := controllerutil.SetControllerReference(m.Controller, clone, m.reconciler.Scheme); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("cannot adopt resource: %s", objType)
 	}
-
-	cmClone := cm.DeepCopy()
-	// Note that ValidateOwnerReferences() will reject this patch if another
-	// OwnerReference exists with controller=true.
-	if err := controllerutil.SetControllerReference(m.Controller, cmClone, m.reconciler.Scheme); err != nil {
-		return err
+	annotations := clone.GetAnnotations()
```

Comment:
Please move the annotation related code to only  statefulset case. And also put the reason behind this logic in comment.